### PR TITLE
Add REX LR Scheduler

### PR DESCRIPTION
**Pull request by xzuyn**
# Description

https://arxiv.org/abs/2107.04197
https://github.com/IvanVassi/REX_LR

```yaml
lr_scheduler: rex
```

Works with `cosine_min_lr_ratio` and `warmup_steps`.

## Screenshots (if appropriate)

Screenshot 2025-03-03 at 11-16-03 LLaMa-3 2-1B-Instruct Workspace – Weights   Biases

<details>
  <summary>Axolotl Config</summary>

```yaml
# Weights and Biases logging config
wandb_project: LLaMa-3.2-1B-Instruct
wandb_entity:
wandb_watch:
wandb_name: rex
wandb_log_model:

# Model checkpointing config
output_dir: ./Outputs/rex
resume_from_checkpoint:
save_steps: 50
save_safetensors: true
save_total_limit: 3
save_only_model: false

# Model architecture config
base_model: meta-llama/Llama-3.2-1B-Instruct
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

# Mixed precision training config
bf16: true
fp16: false
tf32: false

# Model loading config
load_in_8bit: false
load_in_4bit: true
strict: false

# Sequence config
sequence_len: 4096
min_sample_len: 256
sample_packing: true
eval_sample_packing: true
pad_to_sequence_len: true
train_on_inputs: false
group_by_length: false

# LoRA adapter config
adapter: qlora
lora_model_dir:
lora_r: 32
lora_alpha: 32
lora_dropout: 0.125
peft_layers_to_transform:
peft_use_dora:
peft_use_rslora:
peft_layer_replication:
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj
lora_modules_to_save:

# Fix uninitialized tokens (such as <|start_header_id|> on the base L3 models)
fix_untrained_tokens:

# Dataset config
# Completion: https://github.com/xzuyn/axolotl/blob/prompt_formats/src/axolotl/prompt_strategies/customcompletion-regex.py
datasets:
# Completion
  - path: PJMixers-Dev/fundusnewsonly
    split: train[:1024]  # Only the first 1024
    type: customcompletion-regex
test_datasets:
# Completion
  - path: PJMixers-Dev/fundusnewsonly
    split: train[-32:]  # Only the last 32
    type: customcompletion-regex
val_set_size: 0
eval_strategy: steps
eval_steps: 1
dataset_prepared_path: ./00-Tokenized-Datasets/testt
shuffle_merged_datasets: true
dataset_processes:

# Training hyperparameters
num_epochs: 1
gradient_accumulation_steps: 1
micro_batch_size: 8
eval_batch_size: 8
warmup_steps: 0
optimizer: paged_adamw_8bit
optim_args:
optim_target_modules:
lr_scheduler: rex
learning_rate: 1e-5
cosine_min_lr_ratio:
loraplus_lr_ratio:
loraplus_lr_embedding:
weight_decay: 0.1
max_grad_norm: 1
logging_steps: 1

# Model optimization
gradient_checkpointing: unsloth
sdp_attention: true
plugins:
  - axolotl.integrations.liger.LigerPlugin
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
cut_cross_entropy: true
liger_rope: true
liger_rms_norm: true
liger_layer_norm: true
liger_glu_activation: true
liger_cross_entropy: false
liger_fused_linear_cross_entropy: false
lora_mlp_kernel: false
lora_qkv_kernel: false
lora_o_kernel: false

# DeepSpeed
deepspeed:

# Garbage Collection
gc_steps: 1

# Debug config
debug: true
seed: 42

# Token config
special_tokens:
  bos_token: "<|begin_of_text|>"
  eos_token: "<|end_of_text|>"
  pad_token: "<|finetune_right_pad_id|>"
tokens:
```

</details>

---

Screenshot 2025-03-03 at 21-32-28 LLaMa-3 2-1B-Instruct Workspace – Weights   Biases

<details>
  <summary>Axolotl Config</summary>

```yaml
# Weights and Biases logging config
wandb_project: LLaMa-3.2-1B-Instruct
wandb_entity:
wandb_watch:
wandb_name: rex3
wandb_log_model:

# Model checkpointing config
output_dir: ./Outputs/rex3
resume_from_checkpoint:
save_steps: 100
save_safetensors: true
save_total_limit: 3
save_only_model: false

# Model architecture config
base_model: meta-llama/Llama-3.2-1B-Instruct
model_type: AutoModelForCausalLM
tokenizer_type: AutoTokenizer

# Mixed precision training config
bf16: true
fp16: false
tf32: false

# Model loading config
load_in_8bit: false
load_in_4bit: false
strict: false

# Sequence config
sequence_len: 4096
min_sample_len: 256
sample_packing: true
eval_sample_packing: true
pad_to_sequence_len: true
train_on_inputs: false
group_by_length: false

# LoRA adapter config
adapter: lora
lora_model_dir:
lora_r: 64
lora_alpha: 64
lora_dropout: 0.125
peft_layers_to_transform:
peft_use_dora:
peft_use_rslora:
peft_layer_replication:
lora_target_modules:
  - gate_proj
  - down_proj
  - up_proj
  - q_proj
  - v_proj
  - k_proj
  - o_proj
lora_modules_to_save:

# Fix uninitialized tokens (such as <|start_header_id|> on the base L3 models)
fix_untrained_tokens:

# Dataset config
# Instruct: https://github.com/xzuyn/axolotl/blob/prompt_formats/src/axolotl/prompt_strategies/customllama3-regex.py
datasets:
# Instruct
  - path: PJMixers-Dev/sequelbox_Tachibana2-DeepSeek-R1-PREVIEW-Shuffled-ShareGPT
    split: train[128:]  # Everything except the first 128
    type: customllama3-regex
test_datasets:
# Instruct
  - path: PJMixers-Dev/sequelbox_Tachibana2-DeepSeek-R1-PREVIEW-Shuffled-ShareGPT
    split: train[:128]  # Only the first 128
    type: customllama3-regex
val_set_size: 0
eval_strategy: steps
eval_steps: 10
dataset_prepared_path: ./00-Tokenized-Datasets/testt2
shuffle_merged_datasets: true
dataset_processes:

# Training hyperparameters
num_epochs: 1
gradient_accumulation_steps: 1
micro_batch_size: 16
eval_batch_size: 16
warmup_steps: 0
optimizer: paged_adamw_8bit
optim_args:
optim_target_modules:
lr_scheduler: cosine
learning_rate: 2e-5
cosine_min_lr_ratio:
loraplus_lr_ratio:
loraplus_lr_embedding:
weight_decay: 0.1
max_grad_norm: 1
logging_steps: 1

# Early Stopping
early_stopping_patience:

# Model optimization
gradient_checkpointing: unsloth
sdp_attention: true
plugins:
  - axolotl.integrations.liger.LigerPlugin
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
cut_cross_entropy: true
liger_rope: true
liger_rms_norm: true
liger_layer_norm: true
liger_glu_activation: true
liger_cross_entropy: false
liger_fused_linear_cross_entropy: false
lora_mlp_kernel: false
lora_qkv_kernel: false
lora_o_kernel: false

# DeepSpeed
deepspeed:

# Garbage Collection
gc_steps: 1

# Debug config
debug: true
seed: 42

# Token config
special_tokens:
  bos_token: "<|begin_of_text|>"
  eos_token: "<|end_of_text|>"
  pad_token: "<|finetune_right_pad_id|>"
tokens:
```

</details>

### CODE REVIEW

File: src/axolotl/core/trainers/base.py [hunk 1]
```diff
@@ -36,6 +36,74 @@
...
+
+    def get_lr(self):
+        # Warmup phase: if defined, increase lr linearly from 0 to max_lr.
+        if 0 <= (self.last_step - 2) <= (self.num_warmup_steps - 1):
+            return [self.max_lr * (self.last_step - 2) / (self.num_warmup_steps - 1)]
+
+        # Post-warmup phase: adjust step relative to the end of warmup.
+        step_after = (self.last_step - self.num_warmup_steps) - 2
+        remaining_steps = (self.total_steps - self.num_warmup_steps) - 1
+
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: Is there a reason we have an offset of 2/1 here?

File: src/axolotl/core/trainers/base.py [hunk 2]
```diff
@@ -36,6 +36,74 @@
...
+        # Avoid LR spiking
+        if step_after >= remaining_steps or step_after == -1 or remaining_steps <= 0:
+            return [self.min_lr]
+
+        mod_iter = step_after % remaining_steps
+        z = (remaining_steps - mod_iter) / remaining_steps
+        return [self.min_lr + (self.max_lr - self.min_lr) * (z / (0.1 + 0.9 * z))]
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: This scheduler might conflict with `lr_groups` feature, since we're returning only single LR for all optimizer param groups.

Torch ref: https://pytorch.org/docs/stable/optim.html#per-parameter-options

Example on LambdaLR scheduler: https://github.com/pytorch/pytorch/blob/1eba9b3aa3c43f86f4a2c807ac8e12c4a7767340/torch/optim/lr_scheduler.py#L289-L389

For warmup, you may want to do sth like: 
```python
warmup_factor = self.last_step / max(1, self.num_warmup_steps - 1)  # adjust offset if needed, and if you want to set base value for num_warmup_steps in init

return [base_lr * warmup_factor for base_lr in self.base_lrs]
``` 

prevent LR spiking:
```python
return [base_lr * (self.min_lr / self.max_lr) for base_lr in self.base_lrs]
```
and the final return 

```python
rex_factor = self.min_lr / self.max_lr + (1.0 - self.min_lr / self.max_lr) * (z / (0.1 + 0.9 * z)) # this is scaled factor by extracting max_lr out then multiplied later below

return [base_lr * rex_factor for base_lr in self.base_lrs]
```

I *think* this could be how it looks but this will need testing to be sure.

**Review by NanoCode012 — COMMENTED** [review]
Thanks for the integration. I left some comments below.

Could we also add one E2E run test with this optimizer? You can ref to `tests/e2e/test_optimizers.py`.

File: src/axolotl/core/trainers/base.py (hunk 1 above)
Comment (xzuyn) [reply src/axolotl/core/trainers/base.py]: I had issues with the scheduler stopping too early and spiking up to the max_lr. This combination was the simplest thing I could find that fixed it. I will try to clean that up.

File: src/axolotl/core/trainers/base.py [hunk 3]
```diff
@@ -36,6 +36,80 @@
...
+            )
+        if num_warmup_steps > total_steps:
+            raise ValueError(
+                f"num_warmup_steps ({num_warmup_steps}) must be less than or equal to total_steps ({total_steps})."
+            )
+
+        self.min_lr = min_lr
+        self.max_lr = max_lr
+        self.total_steps = total_steps - 1
+        self.num_warmup_steps = num_warmup_steps - 1
```
Comment (NanoCode012) [inline src/axolotl/core/trainers/base.py]: Do you think we should set `max(1, num_warmup_steps - 1)` to set minimum of 1 in case warmup is 0/1?

The calculation in `get_lr` could run on `last_step=0` leading to a div by 0.

```python
        if 0 <= 0 <= 0:
             return [
                 base_lr * 0 / 0 # < -- div by 0
                 for base_lr in self.base_lrs
             ]
```

File: src/axolotl/core/trainers/base.py (hunk 1 above)
Comment (xzuyn) [reply src/axolotl/core/trainers/base.py]: The problem was more with me than the code. I was trying to get to it to what I thought was the correct way to do warmup (5 step warmup: step 1 @ `0 lr`, step 5 @ `max_lr`), but that doesn't match what cosine is doing when doing warmup (5 step warmup: step 1 @ `max_lr * (1/5)`, step 5 @ `max_lr * (5/5)`).

I've changed this to match cosine. `self.last_step = last_step - 1` was still needed to prevent it from spiking at the end of training though.

File: src/axolotl/core/trainers/base.py (hunk 3 above)
Comment (xzuyn) [reply src/axolotl/core/trainers/base.py]: I've changed it to `if 1 <= self.last_step <= self.num_warmup_steps`.

**Comment by xzuyn** [comment]
> Could we also add one E2E run test

I'm not familiar with doing that, but I'll give it a shot later.

**Comment by NanoCode012** [comment]
> > Could we also add one E2E run test
> 
> I'm not familiar with doing that, but I'll give it a shot later.

Hey, you can create a file called `tests/e2e/test_schedulers.py`

and have a similar test to the below (just change scheduler instead of optimizer)

https://github.com/axolotl-ai-cloud/axolotl/blob/75cbd1530172de318f7847250b75785f83c2007b/tests/e2e/test_optimizers.py#L27-L69

**Comment by winglian** [comment]
@xzuyn It would be great if this also accepted the warmup steps too

**Comment by xzuyn** [comment]
It does accept warmup steps. Configs like this work.
```yaml
lr_scheduler: rex
warmup_steps: 50
cosine_min_lr_ratio: 0.05
```

Unless you mean adding it to the `test_schedulers.py`?

**Review by winglian — APPROVED** [review]
thanks @xzuyn ! Good to ship once the tests finish
